src/request/spaces.rs

```diff
@@ -0,0 +1,426 @@
+use super::{ApiRequest, RequestBuilder};
+use crate::error::ErrorKind;
+use chrono::offset::Utc;
+use chrono::DateTime;
+use failure::Error;
+use hmac::{Hmac, Mac};
+use http::header::{self, HeaderMap, HeaderName, HeaderValue};
+use http::method::Method;
+use hyper::{Body, Request};
+use sha2::{Digest as Sha256Digest, Sha256};
+
+use std::collections::BTreeMap;
+
+type HmacSha256 = Hmac<Sha256>;
+
+lazy_static! {
+    /// Access control list header for resources.
+    pub(crate) static ref ACL_HEADER: HeaderName = HeaderName::from_static("x-amz-acl");
+    /// Date header used in authorization.
+    pub(crate) static ref DATE_HEADER: HeaderName = HeaderName::from_static("x-amz-date");
+    /// Hash header used in authorization.
+    pub(crate) static ref HASH_HEADER: HeaderName = HeaderName::from_static("x-amz-content-sha256");
+    /// Headers that are ignored in spaces request.
+    static ref SPACES_IGNORED_HEADERS: [HeaderName; 4] = [
+        header::AUTHORIZATION, header::CONTENT_LENGTH, header::CONTENT_TYPE, ACL_HEADER.clone()
+    ];
+}
+
+/// HTTP request builder for Spaces API.
+// NOTE: This struct is accessible only within the crate, so we don't have to
+// worry about the fields and methods being public.
+// FIXME: A number of owned strings should be borrows here.
+pub(crate) struct SpacesRequestBuilder {
+    /// Relative path of request URL.
+    pub path: Option<String>,
+    /// Payload for this request.
+    pub payload: Option<Vec<u8>>,
+    /// Headers used in this request.
+    pub headers: HeaderMap,
+    /// Content type of this request.
+    pub content_type: Option<HeaderValue>,
+    /// `User-Agent` header to be set in the request.
+    pub user_agent: Option<String>,
+
+    method: Method,
+    region: String,
+    params: BTreeMap<String, String>,
+    host: String,
+    access_key: String,
+    secret_key: String,
+}
+
+impl SpacesRequestBuilder {
+    /// Spaces hostname.
+    const SPACES_HOST: &'static str = "digitaloceanspaces.com";
+    /// Default user agent (set in request header).
+    const DEFAULT_USER_AGENT: &'static str = "ocean";
```

Comment:
Oh, neat. :) I wonder if we should do this globally.

Reply:
No, definitely not. I was imagining this to be the default (as it aligns with the CLI). The client will have an `user_agent` field similar to the builder. It'll be set it in the builder and that'll override the default user agent.